### DISCUSSION TITLE
Possible feature? When using validation resolver, option to submit original data if validation passes, instead of the resolver's output

### DISCUSSION THREAD

**Discussion by adamkangas** [react-hook-form/react-hook-form#2690] [Q&A] [answered]
Currently in `onSubmit` (`useForm.ts:987`), when using a resolver, we pass `fieldValues` into the resolver, and then overwrite `fieldValues` with the `values` returned by the resolver, before sending them to `onValid` if validation passed.

Is there any harm in an option to prevent the overwrite, and submit the original values? Or might that cause a problem I am not expecting?

Schema Validators like Yup and Joi will sometimes cast your data to other types before attempting validation, which can lead to confusing situations... my form contains dates represented as `YYYY-mm-dd` strings which need to be compared during validation, but `Date` objects are being sent to `onValid`, not strings. The validators are casting them to perform comparisons.

To make things more confusing, Yup and Joi return DIFFERENT Date objects for the same input string. There appears to be a difference in how they treat time zones – the values provided by Yup were not working with my backend, but Joi works fine.

Everything is working now after I've switched from Yup to Joi, but if my form could just submit the original string in the first place I would have saved many hours of debugging :)

**Comment by bluebill1049** [comment]
> Currently in `onSubmit` (`useForm.ts:987`), when using a resolver, we pass `fieldValues` into the resolver, and then overwrite `fieldValues` with the `values` returned by the resolver, before sending them to `onValid` if validation passed.

any chance to share a code snippet
<img width="594" alt="Screen Shot 2020-08-24 at 12 47 54 pm" src="https://user-images.githubusercontent.com/10513364/90998673-08ccfc80-e608-11ea-8ef8-71d38dc4ffa5.png">
?

**Comment by adamkangas** [comment, marked as answer]
![react-hook-form_useForm_ts_at_master_·_react-hook-form_react-hook-form](https://user-images.githubusercontent.com/888010/90998894-7d07a000-e608-11ea-83a1-498e0a2862bc.jpg)

**Reply by bluebill1049** [reply]
oh, we want to give user the flexibility to return values if they are running their own resolver. What's the problem you are having? yup should return the correct values as we feed schema validation with our `fieldsValues`.

**Reply by adamkangas** [reply]
I think the problem is that your resolver can do an excellent job of validation, but you might not trust it to provide the final values to your application.

For the date string `2020-08-24`...

Joi casts it to `Mon Aug 24 2020 10:00:00 GMT+1000 (Australian Eastern Standard Time)`
Yup casts it to `Mon Aug 24 2020 00:00:00 GMT+1000 (Australian Eastern Standard Time)`

These are different! That's a problem!

Let me make a quick CSB to show this....

**Reply by bluebill1049** [reply]
can you put a breakpoint before and after the resolver? If yup is mutated the result, then probably should send an issue there.

**Reply by adamkangas** [reply]
https://codesandbox.io/s/yup-vs-joi-date-cast-piob2

Maybe it is a failing with Yup and not the responsibility of react-hook-form, I just did not anticipate that I could pass validation but then immediately submit bad data to my server without a chance to intercept it 🤷🏼‍♂️  (the yup output causes all of the saved dates on the server to be off by one)

Not a big deal, I'll just use Joi... 

Thanks Bill!

**Reply by bluebill1049** [reply]
that's so strange... 👍

**Reply by adamkangas** [reply]
Turns out this is actually an Axios issue, related to this (and with a similar solution): https://github.com/axios/axios/issues/2284

I modified my app to loop through all form data coming from react-hook-form, casting any Date objects back to strings, before making my http request. Axios will leave the strings alone, whereas it was making strange choices about how to serialize the Date objects created by the reducers.

So I am successfully using the Yup Resolver again, I prefer the syntax to Joi 🙌🏼  Hopefully this helps someone else one day

**Reply by bluebill1049** [reply]
thanks for sharing found! @adamkangas 🙏
